data workspace extension foundation - batch 2

alanrenmsft opened the pull request:
1. allow other extensions to register themselves as project provider and handles the loading of the project provider extensions if needed
1. update sql-database-project extension to be a project provider extension for 'sqlproj' project type
1. read the projects information from workspace file and populate the projects view
1. add refresh feature for the workspace view
1. interfaces and services - Sakshi needs these for her add project work
1. some unit test cases - more tests are on their way
1. make the menu items available in workspace view
![image](https://user-images.githubusercontent.com/13777222/92088277-81a72380-ed81-11ea-95f4-017af7bb8819.png)


in the screenshot below, the project path information are read the workspace file and the sqlproj project type is handled by sql-database-project extension, other extensions can also contribute project types and they will be displayed in this view as well.

![image](https://user-images.githubusercontent.com/13777222/92085176-4efb2c00-ed7d-11ea-8cbf-6f50a6cc5400.png)

current test coverage:
![image](https://user-images.githubusercontent.com/13777222/92085746-1ad43b00-ed7e-11ea-9826-8da75427b3c3.png)

Benjin commented on extensions/data-workspace/src/services/workspaceService.ts:
```diff
@@ -0,0 +1,65 @@
+	async getProjectsInWorkspace(): Promise<string[]> {
+		if (vscode.workspace.workspaceFile) {
+			const projects = <string[]>vscode.workspace.getConfiguration(WorkspaceConfigurationName).get(ProjectsConfigurationName);
+			return projects.map(project => path.isAbsolute(project) ? project : path.join(vscode.workspace.rootPath!, project));
```
Should (can?) absolute pathing be forced as a convention for registered project providers?  I could see switching back and forth - or forgetting to in places - being a source of bugs.

alanrenmsft replied:
this area needs more attention when I introduce the add project to workspace in the future PR, I will make sure it works.

Benjin commented on extensions/data-workspace/src/services/workspaceService.ts:
```diff
@@ -0,0 +1,65 @@
+	async getProjectProvider(projectFilePath: string): Promise<dataworkspace.IProjectProvider | undefined> {
+		await this.ensureProviderExtensionLoaded(path.extname(projectFilePath));
+		const providers = ProjectProviderRegistry.providers;
+		for (let i = 0; i < providers.length; i++) {
+			if (providers[i].supportedProjectTypes.findIndex(pt => projectFilePath.toUpperCase().endsWith(pt.projectFileExtension.toUpperCase())) !== -1) {
+				return providers[i];
+			}
+		}
```
A pair of filter expressions might be cleaner here.

Benjin commented on extensions/data-workspace/src/services/workspaceService.ts:
```diff
@@ -0,0 +1,65 @@
+	}
+
+	async getProjectProvider(projectFilePath: string): Promise<dataworkspace.IProjectProvider | undefined> {
+		await this.ensureProviderExtensionLoaded(path.extname(projectFilePath));
```
What's the behavior if there's no provider registered for a given provider?

alanrenmsft replied:
undefined will be returned and eventually a toast notification will be shown listing the projects with no provider. in workspaceTreeDataProvider

Benjin commented on extensions/data-workspace/src/services/workspaceService.ts:
```diff
@@ -0,0 +1,65 @@
+		return [];
+	}
+
+	async getProjectProvider(projectFilePath: string): Promise<dataworkspace.IProjectProvider | undefined> {
```
If I'm reading the code correctly, this is doing multiple passes through the list of providers that both identify something via a supported project type.  I'm probably being too hopeful here, but is there any way to jump from the vscode extension to the provider so we don't have to scan twice?

alanrenmsft replied:
I've refactored the code a bit and introduced a getProviderByProjectType method on the registry. worst case (the provider extension is not loaded), we still have to call this method twice

Benjin commented on extensions/sql-database-projects/package.json:
```diff
@@ -234,82 +239,82 @@
         {
           "command": "sqlDatabaseProjects.build",
-          "when": "view == sqlDatabaseProjectsView && viewItem == databaseProject.itemType.project",
+          "when": "view =~ /^(sqlDatabaseProjectsView|dataworkspace.views.main)$/ && viewItem == databaseProject.itemType.project",
```
I had no idea we could do this with regex; neat!  Is there any reason to keep sqlDatabaseProjectsView around with the introduction of the dataworkspace view?

alanrenmsft replied:
as we discussed the other day, we will be having these 2 views co-exist for a while.

Benjin commented on extensions/sql-database-projects/src/projectProvider/projectPrivider.ts:
```diff
@@ -0,0 +1,37 @@
+/*---------------------------------------------------------------------------------------------
```
Typo in file name: `projectPr*o*vider`

alanrenmsft replied:
oops!

alanrenmsft commented on extensions/data-workspace/src/common/projectProviderRegistry.ts:
```diff
@@ -0,0 +1,34 @@
+		this._providers.length = 0;
+	}
+
+	validateProvider(provider: dataworkspace.IProjectProvider): void {
+	}
```
this is a placeholder, similarly for the workspace service, I will implement the empty methods and add new methods in future PRs

alanrenmsft commented on extensions/data-workspace/src/services/workspaceService.ts:
```diff
@@ -0,0 +1,65 @@
+		await this.ensureProviderExtensionLoaded();
+		const projectTypes: dataworkspace.IProjectType[] = [];
+		ProjectProviderRegistry.providers.forEach(provider => {
+			projectTypes.push(...provider.supportedProjectTypes);
```
that's what the empty method validateProvider is for :)

alanrenmsft commented on extensions/data-workspace/src/services/workspaceService.ts:
```diff
@@ -0,0 +1,65 @@
+		await this.ensureProviderExtensionLoaded();
+		const projectTypes: dataworkspace.IProjectType[] = [];
+		ProjectProviderRegistry.providers.forEach(provider => {
+			projectTypes.push(...provider.supportedProjectTypes);
```
in this extension, I've been using the file extension without the leading dot. later you can clean up the sql-database-project to use the same.

alanrenmsft commented:
not sure why, but this commit history and files are not correct, and the easiest approach to get back to the clean state is to create a new branch and cherry-pick the commits that i made. 

@Benjin I have addressed all the comments in the new branch and sorry for the inconvenience.
